drake/solvers/Optimization.h

```diff
@@ -108,46 +116,123 @@ namespace Drake {
+  };
+
+  /** QuadraticConstraint
+   * @brief  lb <= .5 x'Qx + b'x <= ub
+   */
+  class QuadraticConstraint : public Constraint {
+  public:
+    template <typename DerivedQ, typename Derivedb>
+    QuadraticConstraint(const VariableList& vars, const Eigen::MatrixBase<DerivedQ>& Q, const Eigen::MatrixBase<Derivedb>& b, double lb, double ub)
+            : Constraint(vars,Vector1d::Constant(lb),Vector1d::Constant(ub)), Q(Q), b(b) {}
+    virtual ~QuadraticConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=.5*x.transpose()*Q*x + b.transpose()*x; }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = .5*x.transpose()*Q.cast<TaylorVarXd>()*x + b.cast<TaylorVarXd>().transpose()*x; };
+
+  private:
+    Eigen::MatrixXd Q;
+    Eigen::VectorXd b;
+  };
+
+  class FunctionConstraint : public Constraint {
+  public:
+    FunctionConstraint(const VariableList& vars, const std::shared_ptr<DifferentiableFunction>& func, size_t num_constraints) : Constraint(vars,num_constraints), func(func) {}
+
+    template <typename DerivedLB, typename DerivedUB>
+    FunctionConstraint(const VariableList& vars, const std::shared_ptr<DifferentiableFunction>& func, const Eigen::MatrixBase<DerivedLB>& lb, const Eigen::MatrixBase<DerivedUB>& ub) : Constraint(vars,lb,ub), func(func) {}
+    virtual ~FunctionConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override {
+      func->eval(x,y);
+    }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override {
+      func->eval(x,y);
+    }
+
+  private:
+    const std::shared_ptr<DifferentiableFunction> func;
   };
 
-  // DifferentiableConstraint, PolynomialConstraint, QuadraticConstraint, ComplementarityConstraint, IntegerConstraint, ...
+  // todo: consider implementing DifferentiableConstraint, TwiceDifferentiableConstraint, PolynomialConstraint, QuadraticConstraint, ComplementarityConstraint, IntegerConstraint, ...
+  /** LinearConstraint
+   * @brief Implements a constraint of the form @f -lb <= Ax <= ub @f
+   */
+  class LinearConstraint : public Constraint {
+  public:
+    LinearConstraint(const VariableList& vars, size_t num_constraints) : Constraint(vars, num_constraints) {}
+    template <typename DerivedLB, typename DerivedUB>
+    LinearConstraint(const VariableList& vars, const Eigen::MatrixBase<DerivedLB>& lb, const Eigen::MatrixBase<DerivedUB>& ub) : Constraint(vars,lb,ub) {};
+    virtual ~LinearConstraint() {}
+
+    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=getMatrix()*x; }
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = getMatrix().cast<TaylorVarXd>()*x; };
```

Comment:
Yep, I confused myself. This is fine. Just made sure using this little test program:
```
#include <iostream>

#define EIGEN_INITIALIZE_MATRICES_BY_NAN
#include <Eigen/Dense>
#include <unsupported/Eigen/AutoDiff>

using namespace Eigen;
using namespace std;

int main () {
  Matrix<double, 1, 1> mat1;
  mat1(0) = 2.0;
  double mat2_val = 3.0;
  VectorXd mat2_derivs(3);
  mat2_derivs << 1.0, 2.0, 3.0;

  // dynamic size
  {
    auto mat1_autodiff_dynamic = mat1.cast<AutoDiffScalar<VectorXd>>();
    cout << "mat1_autodiff_dynamic derivative: " << mat1_autodiff_dynamic(0).derivatives().transpose() << endl;
    Matrix<AutoDiffScalar<VectorXd>, 1, 1> mat2_autodiff_dynamic;
    mat2_autodiff_dynamic(0) = AutoDiffScalar<VectorXd>(mat2_val, mat2_derivs);
    auto mat3_autodiff_dynamic = (mat1_autodiff_dynamic * mat2_autodiff_dynamic).eval();
    cout << "mat3_autodiff_dynamic derivatives: " << mat3_autodiff_dynamic(0).derivatives().transpose() << endl;
  }

  // fixed size
  {
    auto mat1_autodiff_fixed = mat1.cast<AutoDiffScalar<Vector3d>>().eval();
    cout << "mat1_autodiff_fixed derivatives: " << mat1_autodiff_fixed(0).derivatives().transpose() << endl;
    Matrix<AutoDiffScalar<Vector3d>, 1, 1> mat2_autodiff_fixed;
    mat2_autodiff_fixed(0) = AutoDiffScalar<VectorXd>(mat2_val, Vector3d(mat2_derivs));
    auto mat3_autodiff_fixed = (mat1_autodiff_fixed * mat2_autodiff_fixed).eval();
    cout << "mat3_autodiff_fixed derivatives: " << mat3_autodiff_fixed(0).derivatives().transpose() << endl;
  }

  return 0;
}
```

Output:
```
mat1_autodiff_dynamic derivative: 
mat3_autodiff_dynamic derivatives: 2 4 6
mat1_autodiff_fixed derivatives: 0 0 0
mat3_autodiff_fixed derivatives: 2 4 6
```

Using the debugger, I verified that the cast call uses this AutoDiffScalar constructor:
```
    /** Conversion from a scalar constant to an active scalar.
      * The derivatives are set to zero. */
    /*explicit*/ AutoDiffScalar(const Real& value)
      : m_value(value)
    {
      if(m_derivatives.size()>0)
        m_derivatives.setZero();
    }
```

So yeah, it sets the derivatives to zero.